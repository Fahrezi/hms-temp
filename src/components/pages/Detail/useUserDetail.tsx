import { useParams } from 'react-router-dom';

import { getUserById } from '@/services/user.service';

import type { User } from '@/types/user';

const useUserDetail = () => {
  const { user_id } = useParams<{ user_id: string }>();

  const { data, isLoading, error } = getUserById(parseInt(user_id || '0'));

  const userData: User | undefined = data?.data.user;

  return {
    userData,
    isLoading,
    error,
  };
};

export default useUserDetail;
