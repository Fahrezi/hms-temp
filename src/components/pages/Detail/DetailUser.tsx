import Loading from '@/components/ui/Loading';

import useUserDetail from './useUserDetail';

import styles from './Detail.module.scss';

const DetailUser = () => {
  const { userData, isLoading, error } = useUserDetail();

  if (isLoading) {
    return (
      <div className={styles.loading}>
        <Loading />
      </div>
    );
  }

  if (error) {
    return <div className={styles.error}>Error: {error.message}</div>;
  }

  if (!userData) {
    return <div className={styles.notFound}>User not found</div>;
  }

  return (
    <div className={styles.userDetail}>
      <h1 onClick={() => window.history.back()}>{'<'} back</h1>
      <h1 className={styles.title}>User Details</h1>
      <div className={styles.card}>
        <div className={styles.field}>
          <span className={styles.label}>ID:</span>
          <span className={styles.value}>{userData.user_id}</span>
        </div>
        <div className={styles.field}>
          <span className={styles.label}>Email:</span>
          <span className={styles.value}>{userData.email}</span>
        </div>
      </div>
    </div>
  );
};

export default DetailUser;
