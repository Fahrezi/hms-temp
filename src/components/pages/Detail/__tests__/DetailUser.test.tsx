/* eslint-disable sonarjs/no-duplicate-string */
import { fireEvent, render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import DetailUser from '@/components/pages/Detail';
import useUserDetail from '@/components/pages/Detail/useUserDetail';

vi.mock('@/components/pages/Detail/useUserDetail');

describe('DetailUser Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders loading state when data is fetching', () => {
    vi.mocked(useUserDetail).mockReturnValue({
      isLoading: true,
      userData: undefined,
      error: null,
    });

    render(<DetailUser />);

    expect(screen.getByTestId('loading')).toBeInTheDocument();
  });

  it('renders error state when there is an error', () => {
    vi.mocked(useUserDetail).mockReturnValue({
      isLoading: false,
      error: {
        message: 'Something went wrong',
        name: '',
      },
      userData: undefined,
    });

    render(<DetailUser />);

    expect(screen.getByText('Error: Something went wrong')).toBeInTheDocument();
  });

  it('renders not found state when userData is null', () => {
    vi.mocked(useUserDetail).mockReturnValue({
      isLoading: false,
      error: null,
      userData: undefined,
    });

    render(<DetailUser />);

    expect(screen.getByText('User not found')).toBeInTheDocument();
  });

  it('renders user details when data is available', () => {
    vi.mocked(useUserDetail).mockReturnValue({
      isLoading: false,
      error: null,
      userData: {
        user_id: 1,
        email: 'user@example.com',
        display_name: '',
        roles: [],
      },
    });

    render(<DetailUser />);

    expect(screen.getByText('User Details')).toBeInTheDocument();
    expect(screen.getByText('ID:')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('Email:')).toBeInTheDocument();
    expect(screen.getByText('user@example.com')).toBeInTheDocument();
  });

  it('navigates back when the back button is clicked', () => {
    window.history.back = vi.fn();

    vi.mocked(useUserDetail).mockReturnValue({
      isLoading: false,
      error: null,
      userData: {
        user_id: 1,
        email: 'user@example.com',
        display_name: '',
        roles: [],
      },
    });

    render(<DetailUser />);

    const backButton = screen.getByText('< back');
    fireEvent.click(backButton);

    expect(window.history.back).toHaveBeenCalled();
  });

  it('renders the snapshot DetailUser component', () => {
    vi.mocked(useUserDetail).mockReturnValue({
      isLoading: false,
      error: null,
      userData: {
        user_id: 1,
        email: 'user@example.com',
        display_name: '',
        roles: [],
      },
    });

    const { container } = render(<DetailUser />);
    expect(container).toMatchSnapshot();
  });
});
