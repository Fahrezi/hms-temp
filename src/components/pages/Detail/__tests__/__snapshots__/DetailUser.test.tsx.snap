// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailUser Component > renders the snapshot DetailUser component 1`] = `
<div>
  <div
    class="_userDetail_75ff1b"
  >
    <h1>
      &lt;
       back
    </h1>
    <h1
      class="_title_75ff1b"
    >
      User Details
    </h1>
    <div
      class="_card_75ff1b"
    >
      <div
        class="_field_75ff1b"
      >
        <span
          class="_label_75ff1b"
        >
          ID:
        </span>
        <span
          class="_value_75ff1b"
        >
          1
        </span>
      </div>
      <div
        class="_field_75ff1b"
      >
        <span
          class="_label_75ff1b"
        >
          Email:
        </span>
        <span
          class="_value_75ff1b"
        >
          user@example.com
        </span>
      </div>
    </div>
  </div>
</div>
`;
